flashlight/fl/tensor/backend/jit/ir/Node.h

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <list>
+#include <optional>
+#include <vector>
+
+#include "flashlight/fl/tensor/TensorBase.h"
+#include "flashlight/fl/tensor/backend/jit/ir/NodeType.h"
+#include "flashlight/fl/tensor/backend/jit/ir/Use.h"
+
+namespace fl {
+
+using UseList = std::list<Use*>;
+
+/**
+ * A Node represents a step in some Tensor computation.
+ *
+ * Conceptually, nodes form a computation DAG where the result of the
+ * computation is immutable (this must be enforced by all graph rewrites).
+ *
+ * Ownership model:
+ * 1. Node has shared ownership via manual refcount update.
+ * 2. Use is owned by user node.
+ */
+class Node {
+  std::vector<Node*> inputs_;
+  std::vector<UseList::iterator> inputUseIters_;
+  UseList uses_;
+  unsigned refCount_{0};
+
+  // present if this node has been evaluated
+  std::optional<Tensor> result_{std::nullopt};
+
+  void nodeImplTypeCheck(NodeType expect, NodeType actual) const;
+
+  // set the input at `inputIdx` -- `resetInput` should be used to clear the old
+  // input at `inputIdx`, if any.
+  void setInputImpl(unsigned inputIdx, Node* input);
+
+  // "unlink" the input at `inputIdx`
+  void resetInput(unsigned inputIdx);
+
+ protected:
+  // A constructor that sets up all the metadata
+  Node(std::vector<Node*>&& inputs);
+
+  // Help enforce internal consistency.
+  Node* getInput(unsigned inputIdx) const;
+
+ public:
+  virtual ~Node();
+
+  // Inputs
+  const std::vector<Node*>& inputs() const;
+  void setInput(unsigned inputIdx, Node* newInput);
+
+  // Uses
+  const UseList& uses() const;
+  void replaceAllUsesWith(Node* newInput);
+
+  // Mainly for debugging/testing
+  unsigned getRefCount() const;
+  // Use carefully -- this manually simulates shared ownership of a node
+  //
+  // Possible stages of lifetime
```

Comment:
Hopefully this makes the refcount & ownership semantics clear.

Essentially, the real owners of a `Node` are the ones who invokes the `incRefCount` method (user Node, JitTensor, GraphVizPrinter). Again, the graph pass writers (the extensible part of the JIT) shouldn't have to deal with this at all.